fix nested admin (one-to-one-to-many relation) that opens inline (fixes #6777)

vladyslavstartsev opened the pull request:
## Subject
Fix form rendering of nested Admin properties.

I am targeting this branch, because everything is backwards compatible.

Closes #6777.

## Changelog

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Fixed
How `AdminHelper::appendFormFieldElement` renders nested forms.
```

vladyslavstartsev commented:
tests will fail, because I have no idea how do you mock methods that does not exist on interface (talking about `hasFormFieldDescription`). yes, I've google it

VincentLanglet commented:
> tests will fail, because I have no idea how do you mock methods that does not exist on interface (talking about `hasFormFieldDescription`). yes, I've google it

```
$this->getMockBuilder(Interface::class)
            ->addMethods(['method'])
            ->getMockForAbstractClass();
```
?

VincentLanglet commented on src/Admin/AdminHelper.php:
```diff
@@ -211,7 +211,7 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
                 $modelClassName = $collection->getTypeClass()->getName();
             } elseif ($collection instanceof Collection) {
-                $modelClassName = $this->getEntityClassName($admin, explode('.', preg_replace('#\[\d*?\]#', '', $path)));
+                $modelClassName = $this->getModelClassName($admin, explode('.', preg_replace('#\[\d*?\]#', '', $path)));
```
Isn't this a BC-break if someone override `getEntityClassName` ?

vladyslavstartsev replied:
I've reverted this one, since it's not relevant to this PR

VincentLanglet commented on src/Admin/AdminHelper.php:
```diff
@@ -202,7 +202,7 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
         //Child form not found (probably nested one)
         //if childFormBuilder was not found resulted in fatal error getName() method call on non object
-        if (!$childFormBuilder) {
+        if (!$childFormBuilder || !$admin->hasFormFieldDescription($childFormBuilder->getName())) {
```
Can you update the comment before to add better explanation ?

vladyslavstartsev replied:
:ok_hand: 

BTW PhpStorm suggested to me to flip `if-else`.
And if will look prettier to me

```php
if ($childFormBuilder && $admin->hasFormFieldDescription($childFormBuilder->getName())) {
``` 

But I think PR should atomic, maybe in next PR

VincentLanglet replied:
It's not a bad idea to flip indead !

vladyslavstartsev replied:
ok, I've flipped it

VincentLanglet reviewed (commented):
Can you add a test where hasFormFieldDescription return false ?

vladyslavstartsev commented:
There isnone case that I dont know how to handle.
There are nested if
```php
            if ($collection instanceof DoctrinePersistentCollection || $collection instanceof PersistentCollection) {
                //since doctrine 2.4
                $modelClassName = $collection->getTypeClass()->getName();
            } elseif ($collection instanceof Collection) {
                $modelClassName = $this->getEntityClassName($admin, explode('.', preg_replace('#\[\d*?\]#', '', $path)));
            } else {
                throw new \Exception('unknown collection class');
            }
```

I can check for exception. But how should I handle case with `PersistentCollection`?
The short answer is - no, just don't check

VincentLanglet commented:
I find the code 
```
if ($collection instanceof DoctrinePersistentCollection || $collection instanceof PersistentCollection) {
                //since doctrine 2.4
                $modelClassName = $collection->getTypeClass()->getName();
            } elseif ($collection instanceof Collection) {
                $modelClassName = $this->getEntityClassName($admin, explode('.', preg_replace('#\[\d*?\]#', '', $path)));
            } else {
                throw new \Exception('unknown collection class');
            }
```
weird, since a PersistentCollection is  a Collection, we should be able to remove the if part.

the collection is 
```
$collection = $this->propertyAccessor->getValue($subject, $path);
```
So you could mock the propertyAccessor.

VincentLanglet commented:
Rebasing 3.x should fix the build

phansys commented on src/Admin/AdminHelper.php:
```diff
@@ -252,6 +230,26 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
 
             $associationAdmin->setSubject($newInstance);
+        } else {
+            //Else branch is executed when child form not found (probably nested one)
+            //if childFormBuilder was not found resulted in fatal error getName() method call on non object
+            //OR if childFormBuilder was found, but admin object does not have formFieldDescription with
+            //same name as childFormBuilder
```
```suggestion
            // The `else` branch is executed when child form was not found (probably nested one).
            // If `$childFormBuilder` was not found resulted in fatal error `getName()` method call on non object
            // or if `$childFormBuilder` was found, but the admin object does not have a formFieldDescription with
            // same name as `$childFormBuilder`.
```

vladyslavstartsev replied:
done

phansys commented on src/Admin/AdminHelper.php:
```diff
@@ -252,6 +230,26 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
+            if ($collection instanceof Collection) {
+                $modelClassName = $this->getEntityClassName($admin, explode('.', preg_replace('#\[\d*?\]#', '', $path)));
+            } else {
+                throw new \Exception('unknown collection class');
```
IMO, we could flip this condition in order to avoid the `else` block.
Regarding the exception, I think we must use a narrower type (like `\TypeError` or similar) and provide a better message explaining what is the expected and the found type.
```suggestion
                throw new \Exception('Unknown collection class.');
```

vladyslavstartsev replied:
Did something that you asked for

phansys replied:
Thank you!

vladyslavstartsev commented:
some why on my local machine `php-cs-fixer` was trying to fix `E_USER_DEPRECATED` in opposite way

```diff
-        ), \E_USER_DEPRECATED);
+        ), E_USER_DEPRECATED);
```

I don't really understand why

VincentLanglet commented:
Can you try `composer update` ? or a rebase ?

vladyslavstartsev commented:
> Can you try `composer update` ? or a rebase ?

1. deleted cache
2. rebased to latest commit on 3.x
3. composer update
still the same.

Any way, I hope that it's been final fixes of this branch, other way I'll fix it some how

phansys commented on src/Admin/AdminHelper.php:
```diff
@@ -252,6 +230,33 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
+
+            $collection = $this->propertyAccessor->getValue($subject, $path);
+
+            if (!($collection instanceof Collection)) {
+                throw new \TypeError(sprintf(
+                    'Unknown collection class. Expected instance of %s, got %s',
+                    Collection::class,
+                    \is_object($collection) ? \get_class($collection) : \gettype($collection)
```
```suggestion
            if (!$collection instanceof Collection) {
                throw new \TypeError(sprintf(
                    'Collection must be an instance of %s, %s given.',
                    Collection::class,
                    \is_object($collection) ? 'instance of "'.\get_class($collection).'"' : '"'.\gettype($collection).'"'
```

vladyslavstartsev replied:
done

phansys commented:
Could you please use the PR template in order to expose the motivation for these changes in the "Changelog" section?
Thank you.

phansys commented on src/Admin/AdminHelper.php:
```diff
@@ -252,6 +230,33 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
+
+            $collection = $this->propertyAccessor->getValue($subject, $path);
+
+            if (!($collection instanceof Collection)) {
```
IMO, we should avoid superfluous parentheses.

vladyslavstartsev replied:
Can we discuss this in dev-kit? 
So it's applied globally

phansys reviewed (approved):
If this PR is fixing something, the "Changelog" section should include the "Fixed" item.

OskarStark commented:
Thank you
